From a86aba45da0feceaa1544eb1304f87c696f3458d Mon Sep 17 00:00:00 2001
From: "max.ma" <max.ma@starfivetech.com>
Date: Sun, 7 Apr 2024 16:49:58 -0700
Subject: [PATCH 31/31] ira: Increase the cost of inner loop spill

---
 gcc/ira-costs.cc | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/gcc/ira-costs.cc b/gcc/ira-costs.cc
index bdb1356af91..06e83366afb 100644
--- a/gcc/ira-costs.cc
+++ b/gcc/ira-costs.cc
@@ -35,7 +35,7 @@ along with GCC; see the file COPYING3.  If not see
 #include "addresses.h"
 #include "reload.h"
 #include "print-rtl.h"
-
+#include "cfgloop.h"
 /* The flags is set up every time when we calculate pseudo register
    classes through function ira_set_pseudo_classes.  */
 static bool pseudo_classes_defined_p = false;
@@ -1730,10 +1730,19 @@ static void
 process_bb_for_costs (basic_block bb)
 {
   rtx_insn *insn;
+  int depth;
 
   frequency = REG_FREQ_FROM_BB (bb);
   if (frequency == 0)
     frequency = 1;
+  if (bb->loop_father)
+    {
+      depth = loop_depth(bb->loop_father);
+      depth = depth > 5 ? 5 : depth;
+      while (depth--)
+        frequency *= 3;
+    }
+
   FOR_BB_INSNS (bb, insn)
     insn = scan_one_insn (insn);
 }
-- 
2.25.1

